phpstan.neon.dist

```diff
@@ -0,0 +1,39 @@
+parameters:
+    level: 3
+    paths:
+        - %currentWorkingDirectory%/lib
+    autoload_files:
+        - %currentWorkingDirectory%/tests/phpstan-polyfill.php
+    reportUnmatchedIgnoredErrors: false
+    ignoreErrors:
+        # extension not available
+        - '~^(Used )?(Function|Constant) sasql_\S+ not found\.\z~i'
+
+        # removing it would be BC break
+        - '~^Constructor of class Doctrine\\DBAL\\Schema\\Table has an unused parameter \$idGeneratorType\.\z~'
+
+        # changing this would be a BC break
+        - '~^Method Doctrine\\DBAL\\Driver\\IBMDB2\\DB2Connection::exec\(\) should return int but returns bool\.\z~'
+        - '~^Method Doctrine\\DBAL\\Query\\QueryBuilder::execute\(\) should return Doctrine\\DBAL\\Driver\\Statement\|int but returns Doctrine\\DBAL\\Driver\\ResultStatement\.\z~'
+
+        # http://php.net/manual/en/pdo.sqlitecreatefunction.php
+        - '~^Call to an undefined method Doctrine\\DBAL\\Driver\\PDOConnection::sqliteCreateFunction\(\)\.\z~'
+
+        # legacy variadic-like signature
+        - '~^Method Doctrine\\DBAL\\Driver\\Connection::query\(\) invoked with \d+ parameters?, 0 required\.\z~'
+
+        # some drivers actually do accept 2nd parameter...
+        - '~^Method Doctrine\\DBAL\\Platforms\\AbstractPlatform::getListTableForeignKeysSQL\(\) invoked with \d+ parameters, 1 required\.\z~'
+
+        # this method is undeclared, relies on PDO ancestor in implementors
+        - '~^Call to an undefined method Doctrine\\DBAL\\Driver\\Statement::nextRowset\(\)\.\z~'
+
+        # legacy remnants from doctrine/common
+        - '~^Class Doctrine\\Common\\(Collections\\Collection|Persistence\\Proxy) not found\.\z~'
+        - '~^.+ on an unknown class Doctrine\\Common\\(Collections\\Collection|Persistence\\Proxy)\.\z~'
+
+        # inheritance variance inference issue
+        - '~^Method Doctrine\\DBAL\\Driver\\PDOConnection::\w+\(\) should return Doctrine\\DBAL\\Driver\\Statement but returns PDOStatement\.\z~'
+
+        # may not exist when pdo_sqlsrv is not loaded
+        - '~^Access to undefined constant PDO::SQLSRV_ENCODING_BINARY\.\z~'
```

Comment:
Could we generate a stub in the longer term?

Reply:
Not sure if possible for a class constant...

Reply:
It's missing from the PDO stub as well as many other driver-specific symbols. I'll file a PR for [JetBrains/phpstorm-stubs](https://github.com/JetBrains/phpstorm-stubs) later.

Reply:
Filed https://github.com/JetBrains/phpstorm-stubs/pull/381.

Reply:
> Not sure if possible for a class constant…

Oh, I see what you're saying. The class stub will be not loaded for PDO in the extension itself is loaded, even if it's compiled w/o the missing constants. PDO is so nicely designed.

Reply:
It's definitely missing in the stubs, but it is also missing in PHP itself unless oci extension is loaded. Who said multiple inheritance in PHP is not possible? :grin: